### DISCUSSION TITLE
must supply `account` to `estimateGas` _even_ when `walletClient` is supplied

### DISCUSSION THREAD

**Discussion by firnprotocol** [wevm/viem#546] [General]
viem (`0.3.31` is requiring that i explicitly supply an `account` into a contract's instance's `estimateGas` action. moreover, the supplied `account` must have enough funds to pay the gas. finally, this is true _even_ when a (funded) `walletClient` has been supplied to the contract. the reason appears to be that `estimateGas` is attached to the `publicClient` supplied, not to the `walletClient`.

after initializing:
```javascript
const account = privateKeyToAccount("0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"); // replace this with a funded account
const publicClient = createPublicClient({
  chain: mainnet,
  transport: http(),
});
const walletClient = createWalletClient({
  account,
  chain: mainnet,
  transport: http(),
});
const contract = getContract({
  address: '0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2',
  abi: wagmiAbi,
  publicClient,
  walletClient,
});
```

compare the following code snippets:
```javascript
const gas = await contract.estimateGas.mint([69420], {
  account,
});
```
the above one works, but _only_ when `account` is sufficiently funded.
```javascript
const gas = await contract.estimateGas.mint([69420]);
```
the above one _never_ works, and throws:
```javascript
ContractFunctionExecutionError: Could not find an Account to execute with this Action.
Please provide an Account with the `account` argument on the Action, or by supplying an `account` to the WalletClient.
```
this error is misleading, since it throws the same error _even when_ an `account` has been supplied to the a `walletClient`, and the `walletClient` itself supplied to the `contract`.

this all seems contrary to the behavior described [here](https://github.com/wagmi-dev/viem/discussions/165).

as a separate point, not sure if it's intended behavior to require that `account` have enough funds to pay the gas, but it seems slightly contrary to the intention of `estimateGas`, which is that we don't know it yet.

not sure that any change is needed, just surfacing a bit of confusing behavior.

**Comment by firnprotocol** [comment]
@tmm FYI? thoughts? thanks in advance for your attention and sorry for the spam.

**Reply by tmm** [reply]
PR up to fall back to wallet client `account` should fix the first error

- https://github.com/wagmi-dev/viem/pull/556

> as a separate point, not sure if it's intended behavior to require that account have enough funds to pay the gas, but it seems slightly contrary to the intention of estimateGas, which is that we don't know it yet.

Worth discussing cc @jxom 

-- 

In the future, please do not tag me directly. If you feel as though something is a bug, feel free to create an issue with a minimal reproduction and we will take a look. Otherwise, I tend to check out discussions as time allows and getting a bunch of notifications reduces signal.

**Reply by firnprotocol** [reply]
gotcha, apologies. won't happen again.

**Comment by firnprotocol** [comment]
hi all, i appear to be still seeing the same behavior in viem `0.3.36`. from [the release description](https://github.com/wagmi-dev/viem/releases/tag/viem%400.3.36), it appears that https://github.com/wagmi-dev/viem/pull/556 should be present in the new release....

partial error dump below:
```javascript
> gas = await contract.estimateGas.mint([69420]);
Uncaught:
ContractFunctionExecutionError: Could not find an Account to execute with this Action.
Please provide an Account with the `account` argument on the Action, or by supplying an `account` to the WalletClient.

Contract Call:
  address:   0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2
  function:  mint(uint256 tokenId)
  args:          (69420)

Docs: https://viem.sh/docs/actions/public/estimateGas.html
Version: viem@0.3.36
    at getContractError (/Users/firnprotocol/relay/node_modules/viem/dist/cjs/utils/errors/getContractError.js:27:12)
    at estimateContractGas (/Users/firnprotocol/relay/node_modules/viem/dist/cjs/actions/public/estimateContractGas.js:24:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async REPL27:1:39 {
  details: undefined,
  docsPath: '/docs/actions/public/estimateGas',
  metaMessages: [
    'Contract Call:',
    '  address:   0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2\n' +
      '  function:  mint(uint256 tokenId)\n' +
      '  args:          (69420)'
  ],
  shortMessage: 'Could not find an Account to execute with this Action.\n' +
    'Please provide an Account with the `account` argument on the Action, or by supplying an `account` to the WalletClient.',
  version: 'viem@0.3.36',
  cause: AccountNotFoundError: Could not find an Account to execute with this Action.
  Please provide an Account with the `account` argument on the Action, or by supplying an `account` to the WalletClient.
```
no clue if i'm doing something wrong...

**Reply by tmm** [reply]
How did you set up `contract`?

**Reply by firnprotocol** [reply]
exactly as in the initial message of the discussion.

full reproducible example:
```javascript
const { createWalletClient, formatUnits, getContract, http, parseGwei, createPublicClient } = require("viem");
var { mainnet } = require('viem/chains');
const { privateKeyToAccount } = require("viem/accounts");

const wagmiAbi = [{"inputs":[],"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"owner","type":"address"},{"indexed":true,"internalType":"address","name":"approved","type":"address"},{"indexed":true,"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"Approval","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"owner","type":"address"},{"indexed":true,"internalType":"address","name":"operator","type":"address"},{"indexed":false,"internalType":"bool","name":"approved","type":"bool"}],"name":"ApprovalForAll","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"from","type":"address"},{"indexed":true,"internalType":"address","name":"to","type":"address"},{"indexed":true,"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"Transfer","type":"event"},{"inputs":[{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"approve","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"owner","type":"address"}],"name":"balanceOf","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"getApproved","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"owner","type":"address"},{"internalType":"address","name":"operator","type":"address"}],"name":"isApprovedForAll","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"mint","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"mint","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[],"name":"name","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"ownerOf","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"safeTransferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"},{"internalType":"bytes","name":"_data","type":"bytes"}],"name":"safeTransferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"operator","type":"address"},{"internalType":"bool","name":"approved","type":"bool"}],"name":"setApprovalForAll","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes4","name":"interfaceId","type":"bytes4"}],"name":"supportsInterface","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"symbol","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"tokenURI","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"pure","type":"function"},{"inputs":[],"name":"totalSupply","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"address","name":"from","type":"address"},{"internalType":"address","name":"to","type":"address"},{"internalType":"uint256","name":"tokenId","type":"uint256"}],"name":"transferFrom","outputs":[],"stateMutability":"nonpayable","type":"function"}];

const account = privateKeyToAccount("0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef");
const publicClient = createPublicClient({
  chain: mainnet,
  transport: http(),
});
const walletClient = createWalletClient({
  account,
  chain: mainnet,
  transport: http(),
});
const contract = getContract({
  address: '0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2',
  abi: wagmiAbi,
  publicClient,
  walletClient,
});

const gas = await contract.estimateGas.mint([69420]); // <--- throws
```
please let me know if you're able to reproduce. i tested this example in Node.js v19.8.1.

**Reply by tmm** [reply]
Got it. The issue is `contract.estimateGas` defaults to the public client right now so you'll need to pass in an account explicitly to make it work. Will take a look at the best way to navigate choosing whether or not to use the wallet or public client.

**Reply by firnprotocol** [reply]
but wait—if you simply omit the `publicClient` from the `contract`, then the `estimateGas` method is not even present at all! (only `write` is.) so not sure what you mean by "defaults to". relatedly, i'm suddenly unsure what https://github.com/wagmi-dev/viem/pull/556 actually changed.

**Reply by tmm** [reply]
> if you simply omit the publicClient from the contract, then the estimateGas method is not even present at all!

Yes, because it only uses the public client right now. So if you don't supply a public client, you can't use `contract.estimateGas`. The types for `estimateGas` are wrong right now. They should be requiring `account`.

> i'm suddenly unsure what https://github.com/wagmi-dev/viem/pull/556 actually changed.

#556 changed the behavior of [`estimateGas`](https://viem.sh/docs/actions/public/estimateGas.html).

--

Two things to fix:

1. Fix typing to require `account` when using `contract.estimateGas` with only public client
2. Add support for wallet client `contract.estimateGas` - need to figure out the behavior for when you create a contract with a public client and wallet client - which one is used to call `estimateGas` internally.

Make sense?

**Reply by firnprotocol** [reply]
> Make sense?

not exactly.... i guess what https://github.com/wagmi-dev/viem/pull/556 changed is the behavior of the _standalone method_ `estimateGas` (?). i was not aware that this method even existed—and (as i presume should be clear from my message above) it was not the subject of my initial report. to make things more confusing, it [looks like](https://viem.sh/docs/actions/public/estimateGas.html) there's a further `estimateGas` attached to `publicClient`s—though such clients don't impart the method to `contract`s (?).
> Two things to fix:

as for these two things, for what it's worth, it's only "Add support for wallet client `contract.estimateGas`" that i need. I don't have the need to supply both public and wallet clients to a contract (except insofar as it's necessary to `estimateGas`).
